rcl_lifecycle/src/com_interface.c

```diff
@@ -0,0 +1,203 @@
+      RCL_SET_ERROR_MSG("Topic name exceeds maximum size of 255");
+      com_interface = NULL;
+      return RCL_RET_ERROR;
+    }
+
+    rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+    rcl_ret_t ret = rcl_publisher_init(&com_interface->state_publisher,
+        node_handle, ts_pub_notify, topic_name, &publisher_options);
+    free(topic_name);
+
+    if (ret != RCL_RET_OK) {
+      com_interface = NULL;
+      return ret;
+    }
+
+    // initialize static message for notification
+    lifecycle_msgs__msg__TransitionEvent__init(&msg);
+  }
+
+  {  // initialize get state service
+     // Build topic, topic suffix hardcoded for now
+     // and limited in length of 255
+    const char * topic_prefix = "__get_state";
+    char * topic_name;
+    if (concatenate(&node_name, &topic_prefix, &topic_name) != true) {
+      RCL_SET_ERROR_MSG("Topic name exceeds maximum size of 255");
+      com_interface = NULL;
+      return RCL_RET_ERROR;
+    }
+
+    rcl_service_options_t service_options = rcl_service_get_default_options();
+    rcl_ret_t ret = rcl_service_init(&com_interface->srv_get_state,
+        node_handle, ts_srv_get_state, topic_name, &service_options);
+    free(topic_name);
+
+    if (ret != RCL_RET_OK) {
+      com_interface = NULL;
+      return ret;
+    }
+  }
+
+  {  // initialize change state service
+     // Build topic, topic suffix hardcoded for now
+     // and limited in length of 255
+    const char * topic_prefix = "__change_state";
+    char * topic_name;
+    if (concatenate(&node_name, &topic_prefix, &topic_name) != true) {
+      RCL_SET_ERROR_MSG("Topic name exceeds maximum size of 255");
+      com_interface = NULL;
+      return RCL_RET_ERROR;
+    }
+
+    rcl_service_options_t service_options = rcl_service_get_default_options();
+    rcl_ret_t ret = rcl_service_init(&com_interface->srv_change_state,
+        node_handle, ts_srv_change_state, topic_name, &service_options);
+    free(topic_name);
+
+    if (ret != RCL_RET_OK) {
+      com_interface = NULL;
+      return ret;
+    }
+  }
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_lifecycle_com_interface_fini(rcl_lifecycle_com_interface_t * com_interface,
+  rcl_node_t * node_handle)
+{
+  rcl_ret_t fcn_ret = RCL_RET_OK;
+  {  // destroy get state srv
+    rcl_ret_t ret = rcl_service_fini(&com_interface->srv_get_state,
+        node_handle);
+    if (ret != RCL_RET_OK) {
+      fcn_ret = RCL_RET_ERROR;
+    }
+  }
+
+  {  // destroy change state srv
+    rcl_ret_t ret = rcl_service_fini(&com_interface->srv_change_state,
+        node_handle);
+    if (ret != RCL_RET_OK) {
+      fcn_ret = RCL_RET_ERROR;
+    }
+  }
+
+  {  // destroy the publisher
+    rcl_ret_t ret = rcl_publisher_fini(&com_interface->state_publisher,
+        node_handle);
+    if (ret != RCL_RET_OK) {
+      fcn_ret = RCL_RET_ERROR;
+    }
+    lifecycle_msgs__msg__TransitionEvent__fini(&msg);
```

Comment:
If you were going for reverse order destruction, then this message should be finalized before the publisher.